src/modules/credential/userpass/path_login.rs

```diff
@@ -0,0 +1,78 @@
+use std::{collections::HashMap, sync::Arc};
+
+use super::{UserPassBackend, UserPassBackendInner};
+use crate::{
+    errors::RvError,
+    logical::{Auth, Backend, Field, FieldType, Lease, Operation, Path, PathOperation, Request, Response},
+    new_fields, new_fields_internal, new_path, new_path_internal,
+};
+
+impl UserPassBackend {
+    pub fn login_path(&self) -> Path {
+        let userpass_backend_ref = Arc::clone(&self.inner);
+
+        let path = new_path!({
+            pattern: r"login/(?P<username>\w[\w-]+\w)",
+            fields: {
+                "username": {
+                    field_type: FieldType::Str,
+                    required: true,
+                    description: "Username of the user."
+                },
+                "password": {
+                    field_type: FieldType::Str,
+                    required: true,
+                    description: "Password for this user."
+                }
+            },
+            operations: [
+                {op: Operation::Write, handler: userpass_backend_ref.login}
+            ],
+            help: r#"This endpoint authenticates using a username and password."#
+        });
+
+        path
+    }
+}
+
+impl UserPassBackendInner {
+    pub fn login(&self, _backend: &dyn Backend, req: &mut Request) -> Result<Option<Response>, RvError> {
+        let err_info = "invalid username or password";
+        let username_vale = req.get_data("username")?;
+        let username = username_vale.as_str().unwrap().to_lowercase();
+        let password_vale = req.get_data("password")?;
```

Comment:
If `password_vale` is cloned from req, then it also needs to be auto-zeroized when its life ends.

Reply:
BTW, `password_vale` is a typo...

Reply:
> BTW, `password_vale` is a typo...

fixed